src/eval/operation.rs

```diff
@@ -1119,17 +1124,33 @@ impl<R: ImportResolver, C: Cache> VirtualMachine<R, C> {
                         Term::Record(record) if !record.fields.is_empty() => {
                             let mut shared_env = Environment::new();
 
-                            let record = record.map_values_closurize(&mut self.cache, &mut shared_env, &env, |_, t| {
-                                mk_term::op1(UnaryOp::Force(), t)
-                            }).map_err(|missing_field_err| missing_field_err.into_eval_err(pos, pos_op))?;
+                            let fields = record.fields
+                                .into_iter()
+                                .filter(|(_, field)| {
+                                    !(field.is_empty_optional() || (for_export && field.metadata.not_exported))
+                                })
+                                .map_values_closurize(&mut self.cache, &mut shared_env, &env, |_, value| {
+                                    mk_term::op1(UnaryOp::Force { for_export }, value)
+                                })
+                                .map_err(|e| e.into_eval_err(pos, pos_op))?;
+
+                            let terms = fields
+                                .values()
+                                .cloned()
+                                .map(|field| {
+                                    field
+                                        .value
+                                        .expect("fields without definition would have thrown MissingFieldDefError before")
+                                })
+                                .collect::<Vec<_>>();
```

Comment:
This `collect` should actually be superfluous.
```suggestion
                                });
```